patterns/direct_processor.py

```diff
@@ -0,0 +1,142 @@
+
+
+if platform.uname().system == 'Windows': os.environ["PYTHONUNBUFFERED"] = "1" # Disable stdout buffer 
+
+# Goal is to demonstrate receive-process-reply (processor) pattern. Processor subscribes to a request topic and upon receipt of 
+# request message carrying a string, the processor converts the payload to uppercase (for simplicity) and publishes to a 
+# designated topic.
+#   direct_requestor: 
+#       subscribes to request topic 'solace/samples/python/direct/processor/input/{count}
+#       proceses the payload
+#       publishes to reply topic 'solace/samples/python/direct/processor/output'
+
+TOPIC_PREFIX = "solace/samples/python"
+
+# Broker Config. Note: Could pass other properties Look into
+broker_props = {
+    "solace.messaging.transport.host": os.environ.get('SOLACE_HOST') or "tcp://localhost:55555,tcp://localhost:55554",
+    "solace.messaging.service.vpn-name": os.environ.get('SOLACE_VPN') or "default",
+    "solace.messaging.authentication.scheme.basic.username": os.environ.get('SOLACE_USERNAME') or "default",
+    "solace.messaging.authentication.scheme.basic.password": os.environ.get('SOLACE_PASSWORD') or ""
+    }
+
+# Build A messaging service with a reconnection strategy of 20 retries over an interval of 3 seconds
+# Note: The reconnections strategy could also be configured using the broker properties object
+messaging_service = MessagingService.builder().from_properties(broker_props)\
+                        .with_reconnection_retry_strategy(RetryStrategy.parametrized_retry(20,3))\
+                        .build()
+
+# Blocking connect thread
+messaging_service.connect()
+print(f'\nMessaging Service connected? {messaging_service.is_connected}')
+
+# Inner classes for connection error handling
+class ServiceEventHandler(ReconnectionListener, ReconnectionAttemptListener, ServiceInterruptionListener):
+    def on_reconnected(self, e: ServiceEvent):
+        print("\non_reconnected")
+        print(f"Error cause: {e.get_cause()}")
+        print(f"Message: {e.get_message()}")
+    
+    def on_reconnecting(self, e: "ServiceEvent"):
+        print("\non_reconnecting")
+        print(f"Error cause: {e.get_cause()}")
+        print(f"Message: {e.get_message()}")
+
+    def on_service_interrupted(self, e: "ServiceEvent"):
+        print("\non_service_interrupted")
+        print(f"Error cause: {e.get_cause()}")
+        print(f"Message: {e.get_message()}")
+
+# Event Handling for the messaging service
+service_handler = ServiceEventHandler()
+messaging_service.add_reconnection_listener(service_handler)
+messaging_service.add_reconnection_attempt_listener(service_handler)
+messaging_service.add_service_interruption_listener(service_handler)
+ 
+# Inner classes for publish error handling
+class PublisherErrorHandling(PublishFailureListener):
+    def on_failed_publish(self, e: "FailedPublishEvent"):
+        print("on_failed_publish")
+
+# Handle received messages
+class MessageHandlerImpl(MessageHandler):
```

Comment:
Open to suggestions here: I would recommend we call this `ProcessorImpl` and assigning it as follows `direct_receiver.receive_async(ProcessorImpl())` and add descriptive comment before the start of the class. Thoughts?

Reply:
I would also declare the publisher in this class as well 

```
def __init__(self, publisher):
        self.publisher = publisher
```
and you can pass it to the class as follows
`direct_receiver.receive_async(MessageHandlerImpl(direct_publisher))`.

This is reflective of good practice incase the MessageHandlerImpl (or ProcessorImpl) is moved to another file and imported to the current file

Reply:
Done!

Reply:
I guess we are keeping all the inner classes in the same file across the samples - for consistency, let us leave it as it is. When we do the next round of house-cleaning, we will organize the inner classes in separate files and explore reusability. This is something you suggested as well :)